### DISCUSSION TITLE
ESP32S3: PCNT: add support for all counting units

### DISCUSSION THREAD

**Discussion by Hallootto1100** [zephyrproject-rtos/zephyr#72933] [RFC]
# ESP32: PCNT: add support for all counting units
## Overview
The `ESP32-S3` features a `PCNT` (Pulse Counter) peripheral containing 4 independent units, each with two channels, making it capable of interfacing with up to 4 quadrature encoders simultaneously.

## Current Status
I successfully configured the first unit and tested it with the [qdec](https://docs.zephyrproject.org/latest/samples/sensor/qdec/README.html) sample. While it's possible to define additional units in the device tree, no new device instances are created. Examining the [current driver implementation](https://github.com/zephyrproject-rtos/zephyr/blob/a2ebe49259d12c7c3375669f65ed5703ddd7f33a/drivers/sensor/espressif/pcnt_esp32/pcnt_esp32.c#L111) reveals that it hardcodes channel 0, effectively limiting functionality to a single unit.

Additionally, the driver implements `SENSOR_CHAN_ROTATION` through the sensor API without performing any unit conversion to degrees, simply returning raw counter values.

## Root Cause
The fundamental limitation stems from the traditional sensor API's inability to support multiple channels of the same type within a single device. However, Zephyr's new [Read and Decode API](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor/read_and_decode.html) addresses this constraint by:

- Supporting multiple channels of identical types per device
- Implementing a decoder mechanism that enables two-stage data processing
- Providing flexibility to access either raw values or decoded measurements (angles, distances, etc.)

Although the Read and Decode API remains experimental, it's actively developed and approaching stability.

## Proposed Solution:
Develop a new `PCNT` driver implementation that leverages the Read and Decode API's capabilities. The driver selection could be controlled via `CONFIG_SENSOR_ASYNC_API`, allowing users to choose between the legacy implementation (single unit) and the new multi-unit capable driver based on their requirements.

This approach would unlock the `ESP32-S3` `PCNT` peripheral's full potential while maintaining backward compatibility with existing applications.

**Comment by Hallootto1100** [comment]
@sylvioalves could you have a look?

**Comment by yunsmall** [comment]
I like this proposal. I was just about to use multiple encoders and was worried that the PCNT driver of ESP32 was not very user-friendly.

**Comment by kpochwala** [comment]
Hello,

I currently stumbled upon this. Is it possible to use multiple channels? I have a robot with two encoder motors and esp32-s3. At first I was really happy to get one of the encoders going using this snippet in my devicetree file (in fact I was flabbergasted that it worked so easily, Zephyr is great!):

```c
&pinctrl {
    pcnt_default: pcnt_default {
        group1 {
            pinmux = <PCNT0_CH0SIG_GPIO8>,
                 <PCNT0_CH0CTRL_GPIO11>,
            bias-pull-up;
        };
    };
};

&pcnt {
    pinctrl-0 = <&pcnt_default>;
    pinctrl-names = "default";
    status = "okay";
    #address-cells = <1>;
    #size-cells = <0>;

    unit0@0 {
        reg = <0>;
        #address-cells = <1>;
        #size-cells = <0>;
        filter = <100>;

        channelA@0 {
            reg = <0>;
            sig-pos-mode = <2>;
            sig-neg-mode = <1>;
            ctrl-h-mode = <0>;
            ctrl-l-mode = <1>;
        };
    };
};

```

and extended `prj.conf`:

```c
CONFIG_SENSOR=y
CONFIG_SENSOR_SHELL=y
CONFIG_SENSOR_INFO=y
```

I was able to read sensor values using shell. Then I've modified the devicetree for what I've thought would add support to the second encoder:

`pinctrl` section:
```c
&pinctrl {
    pcnt_default: pcnt_default {
        group1 {
            pinmux = <PCNT0_CH0SIG_GPIO8>,
                 <PCNT0_CH0CTRL_GPIO11>,
                 <PCNT1_CH0SIG_GPIO4>,
                 <PCNT1_CH0CTRL_GPIO5>;
            bias-pull-up;
        };
    };
};
```

and `pcnt` section:
```c
&pcnt {
    pinctrl-0 = <&pcnt_default>;
    pinctrl-names = "default";
    status = "okay";
    #address-cells = <1>;
    #size-cells = <0>;

    unit0@0 {
        reg = <0>;
        #address-cells = <1>;
        #size-cells = <0>;
        filter = <100>;

        channelA@0 {
            reg = <0>;
            sig-pos-mode = <2>;
            sig-neg-mode = <1>;
            ctrl-h-mode = <0>;
            ctrl-l-mode = <1>;
        };
    };

    unit1@1 {
        reg = <1>;
        #address-cells = <1>;
        #size-cells = <0>;
        filter = <100>;

        channelA@0 {
            reg = <0>;
            sig-pos-mode = <2>;
            sig-neg-mode = <1>;
            ctrl-h-mode = <0>;
            ctrl-l-mode = <1>;
        };
    };
};
```

However I'm getting single channel using shell:

<img width="773" height="152" alt="image" src="https://github.com/user-attachments/assets/595168b6-dbcb-4d65-b88d-132806318594" />


What can I do to achieve reading from both channels?

**Comment by sylvioalves** [comment]
Folks, let me know if https://github.com/zephyrproject-rtos/zephyr/pull/107683 solves it.

**Comment by kpochwala** [comment]
@sylvioalves I've tested code from Your branch - `qdec_multi` sample tweaked to my hardware and it works - thanks a lot! Looking at the timeline of posts in this RFC I must admit that I totally did not expect proof of concept, let alone modern version so quickly.
